packages/aws-cdk-lib/aws-stepfunctions/lib/util.ts

```diff
@@ -1,36 +1,20 @@
-import { IKey } from '../../aws-kms';
-import { Duration } from '../../core';
+import { EncryptionConfiguration } from './encryptionconfiguration';
 
 export function noEmptyObject<A>(o: Record<string, A>): Record<string, A> | undefined {
   if (Object.keys(o).length === 0) { return undefined; }
   return o;
 }
 
-function isInValidKmsDataKeyReusePeriodSeconds(kmsDataKeyReusePeriodSeconds: Duration) {
-  return kmsDataKeyReusePeriodSeconds.toSeconds() < 60 || kmsDataKeyReusePeriodSeconds.toSeconds() > 900;
-}
-
-function validateEncryptionConfiguration(kmsKey: IKey | undefined, kmsDataKeyReusePeriodSeconds: Duration | undefined) {
-  if (!kmsKey && kmsDataKeyReusePeriodSeconds) {
-    throw new Error('You cannot set kmsDataKeyReusePeriodSeconds without providing a value for kmsKey');
-  }
-  if (kmsKey && kmsDataKeyReusePeriodSeconds && isInValidKmsDataKeyReusePeriodSeconds(kmsDataKeyReusePeriodSeconds)) {
-    throw new Error('kmsDataKeyReusePeriodSeconds must have a value between 60 and 900 seconds');
-  }
-}
-
-export function constructEncryptionConfiguration(kmsKey: IKey | undefined, kmsDataKeyReusePeriodSeconds: Duration | undefined) {
-  validateEncryptionConfiguration(kmsKey, kmsDataKeyReusePeriodSeconds);
-
-  if (!kmsKey) {
+export function constructEncryptionConfiguration(encryptionConfiguration? : EncryptionConfiguration) {
```

Comment:
Can move to encryptionconfiguration module and specify return type

https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_stepfunctions.CfnStateMachine.EncryptionConfigurationProperty.html

https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_stepfunctions.CfnActivity.EncryptionConfigurationProperty.html

```ts
export function buildEncryptionConfiguration(encryptionConfiguration?: EncryptionConfiguration): CfnStateMachine.EncryptionConfigurationProperty | CfnActivity.EncryptionConfigurationProperty | undefined {
```

"build" instead of "construct" for similarity with `buildLoggingConfiguration` and `buildTracingConfiguration` in state-machine.ts

Reply:
I realized that the current implementation doesn't allow customers to go back to `AWS_OWNED_KEY` after setting `CUSTOMER_MANAGED_KEY`. In order for to customers to change encryption configuration from either `CUSTOMER_MANAGED_KEY` to `AWS_OWNED_KEY` or vice versa, they need to provide a **non-null** encryptionConfiguration when updating StateMachines. I verified by calling the API itself via cli (again this is not something that's documented).

### Proposed solution
I'm proposing we add another field to the encryptionConfiguration module and call it `type` (or even encryption type). This will be a mandatory field when creating an object of type EncryptionConfiguration. Customers can then specify either `AWS_OWNED_KEY` or `CUSTOMER_MANAGED_KMS_KEY` when constructing encryptionConfiguration.

``` ts
import { CfnActivity, CfnStateMachine } from './stepfunctions.generated';
import * as kms from '../../aws-kms';
import * as cdk from '../../core';

/**
 * What kind of server-side encryption to apply to either a StateMachine or Activity
 */
export enum EncryptionType {
  /**
   * Server-side encryption of the state machine definition using an AAWS owned key
   */
  AWS_OWNED_KEY = 'AWS_OWNED_KEY',

  /**
   * Server-side KMS encryption with a customer managed key.
   */
  CUSTOMER_MANAGED_KMS_KEY = 'CUSTOMER_MANAGED_KMS_KEY',
}

/**
 * Define a new EncryptionConfiguration
 */
export class EncryptionConfiguration {
  /**
   * The symmetric customer managed KMS key for server-side encryption of the state machine definition, and execution history or activity inputs.
   * Step Functions will reuse the key for a maximum of `kmsDataKeyReusePeriodSeconds`.
   *
   * @default - data is transparently encrypted using an AWS owned key
   */
  public readonly kmsKey?: kms.IKey;
  /**
   * Maximum duration that Step Functions will reuse customer managed data keys.
   * When the period expires, Step Functions will call GenerateDataKey.
   *
   * Must be between 60 and 900 seconds.
   *
   * @default Duration.seconds(300)
   */
  public readonly kmsDataKeyReusePeriodSeconds?: cdk.Duration;

  /**
   * The encryption option specified for the state machine or activity
  */
  public readonly type: EncryptionType;
  constructor(type: EncryptionType, kmsKey?: kms.IKey, kmsDataKeyReusePeriodSeconds?: cdk.Duration) {
    this.type = type;
    this.validateEncryptionConfiguration(type, kmsKey, kmsDataKeyReusePeriodSeconds);
    this.kmsKey = kmsKey;
    this.kmsDataKeyReusePeriodSeconds = kmsDataKeyReusePeriodSeconds;
  }

  private isInvalidKmsDataKeyReusePeriodSeconds(kmsDataKeyReusePeriodSeconds: cdk.Duration) {
    return kmsDataKeyReusePeriodSeconds.toSeconds() < 60 || kmsDataKeyReusePeriodSeconds.toSeconds() > 900;
  }

  private validateEncryptionConfiguration(type: EncryptionType, kmsKey?: kms.IKey, kmsDataKeyReusePeriodSeconds?: cdk.Duration) {
    if (type === EncryptionType.AWS_OWNED_KEY && (kmsKey || kmsDataKeyReusePeriodSeconds)) {
      throw new Error('InvalidEncryptionConfiguration, you cannot specify a KMS key or kmsDataKeyReusePeriodSeconds when using AWS_OWNED_KEY');
    }
    if (type === EncryptionType.CUSTOMER_MANAGED_KMS_KEY && !kmsKey) {
      throw new Error('InvalidEncryptionConfiguration, you cannot specify CUSTOMER_MANAGED_KMS_KEY without providing kmsKey');
    }
    if (type === EncryptionType.CUSTOMER_MANAGED_KMS_KEY &&
      kmsKey && kmsDataKeyReusePeriodSeconds && this.isInvalidKmsDataKeyReusePeriodSeconds(kmsDataKeyReusePeriodSeconds)) {
      throw new Error('kmsDataKeyReusePeriodSeconds must have a value between 60 and 900 seconds');
    }
  }
}

// eslint-disable-next-line max-len
export function buildEncryptionConfiguration(encryptionConfiguration?: EncryptionConfiguration): CfnStateMachine.EncryptionConfigurationProperty | CfnActivity.EncryptionConfigurationProperty | undefined {
  if (!encryptionConfiguration) {
    return undefined;
  }

  if (encryptionConfiguration.type === EncryptionType.AWS_OWNED_KEY) {
    return {
      type: EncryptionType.AWS_OWNED_KEY,
    };
  }

  // Default value for `kmsDataKeyReusePeriodSeconds`, see: https://docs.aws.amazon.com/step-functions/latest/dg/encryption-at-rest.html#cfn-resources-for-encryption-configuration
  const DEFAULT_KMS_DATA_KEY_REUSE_PERIOD_SECONDS = 300;
  return {
    kmsKeyId: encryptionConfiguration.kmsKey!.keyArn,
    kmsDataKeyReusePeriodSeconds: encryptionConfiguration.kmsDataKeyReusePeriodSeconds ?
      encryptionConfiguration.kmsDataKeyReusePeriodSeconds.toSeconds() : DEFAULT_KMS_DATA_KEY_REUSE_PERIOD_SECONDS,
    type: EncryptionType.CUSTOMER_MANAGED_KMS_KEY,
  };

}
```

#### Using AWS_OWNED_KEY

``` ts
const stateMachine = new sfn.StateMachine(this, 'StateMachineWithCMKEncryptionConfiguration', {
  /* ... other props ... */
  encryptionConfiguration: new EncryptionConfiguration(EncryptionType.AWS_OWNED_KEY)
});
```

#### Using CUSTOMER_MANAGED_KEY (with kmsDataKeyReusePeriodSeconds)

``` ts
const kmsKey = new kms.Key(this, 'Key');
const stateMachine = new sfn.StateMachine(this, 'StateMachineWithCMKEncryptionConfiguration', {
  /* ... other props ... */
  encryptionConfiguration: new EncryptionConfiguration(EncryptionType.CUSTOMER_MANAGED_KMS_KEY, kmsKey, cdk.Duration.seconds(60))
});
```

#### Using CUSTOMER_MANAGED_KEY (without kmsDataKeyReusePeriodSeconds)
``` ts
const kmsKey = new kms.Key(this, 'Key');
const stateMachine = new sfn.StateMachine(this, 'StateMachineWithCMKEncryptionConfiguration', {
  /* ... other props ... */
  encryptionConfiguration: new EncryptionConfiguration(EncryptionType.CUSTOMER_MANAGED_KMS_KEY, kmsKey)
});
```

Thoughts @paulhcsun @wong-a ?

Reply:
Sounds like a similar issue to https://github.com/aws/aws-cdk/issues/30814

In the UpdateStateMachine API, encryptionConfiguration etc. are optional parameters and not providing a new value uses the existing value. The CFN handler incorrectly does this too — the CFN template represents the state of the resource, not a diff to apply. If you remove a property from the template, the handler passes `null` to the API, which tells Step Functions to use the existing value instead of the default.

Reply:
Proposed solution makes sense to me, though I'd prefer avoiding the additional `type` parameter if possible. 

You will also have to update any references to `encryptionConfiguration` as this can be defined, but not have a kmsKey.

Reply:
Agreed that I don't prefer the use of an extra `type` parameter as it puts the responsibility on the user to make sure their type matches correctly with the Key the pass in. 

What you could do instead is create a second class called `AwsOwnedEncryptionConfiguration` (open to other names) where the constructor does not take in a key and rename the original to `CustomerManagedEncryptionConfiguration`. This way you let the user choose which to use and then there's no need to provide guardrails with the error handling here:
```ts
  private validateEncryptionConfiguration(type: EncryptionType, kmsKey?: kms.IKey, kmsDataKeyReusePeriodSeconds?: cdk.Duration) {
    if (type === EncryptionType.AWS_OWNED_KEY && (kmsKey || kmsDataKeyReusePeriodSeconds)) {
      throw new Error('InvalidEncryptionConfiguration, you cannot specify a KMS key or kmsDataKeyReusePeriodSeconds when using AWS_OWNED_KEY');
    }
    if (type === EncryptionType.CUSTOMER_MANAGED_KMS_KEY && !kmsKey) {
      throw new Error('InvalidEncryptionConfiguration, you cannot specify CUSTOMER_MANAGED_KMS_KEY without providing kmsKey');
    }
...
}
```

You also mentioned that this ability to switch between key types is possible but not documented within SFN docs, would it be possible to also add that documentation and link it somewhere in the readme with a `@see` tag?

Reply:
Will use two classes going forward: 1. `CustomerManagedEncryptionConfiguration` 2. `AwsOwnedEncryptionConfiguration`. Will also update the README to mention that customers need to explicitly provide a non-null encryption configuration object when they want to switch encryption types.

Reply:
Ok I will go ahead and use two separate classes based on the recommendation above: 1. `CustomerManagedEncryptionConfiguration` 2. `AwsOwnedEncryptionConfiguration`. Discusses with paulhcsun offline that the documentation link is non-blocking. I will go ahead and update the README to mention that if customer wants to switch the encryption type, they will need to provide a **non-null** encryption configuration.